packages/flutter/lib/src/widgets/resolution_awareness.dart

```diff
@@ -0,0 +1,127 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import 'dart:collection';
+
+import 'package:flutter/services.dart';
+
+import 'basic.dart';
+import 'framework.dart';
+import 'media_query.dart';
+
+/// Asset resolver that knows how to find the asset most closely matching the
+/// current device pixel ratio, as given by [MediaQueryData].
+///
+/// Resolution aware assets are provided in groups. The group name is given
+/// as the asset key to the resolver, which then determines which specific
+/// asset in the group should be used.
+///
+/// For each group, a metadata asset should exist keyed by the group name +
+/// ".meta". This metadata asset is a textual list of individual assets for
+/// that group, separated by newlines. Individual assets should be keyed with a
+/// prefix of "Nx/" where N is the nominal device pixel ratio for that asset.
+/// N does not need to be an integer.
+///
+/// For example, suppose an application wants to use a logical icon named
+/// "heart". This icon has representations at 1.0, 1.5, and 2.0 pixel ratios. The
+/// asset bundle should then contain the following assets:
+///
+/// heart.meta
+/// 1.0x/heart.png
+/// 1.5x/heart.png
+/// 2.0x/heart.png
+///
+/// On a device with a 1.0 device pixel ratio, the image chosen would be
+/// 1.0x/heart.png; on a device with a 1.3 device pixel ratio, the image chosen
+/// would be 1.5x/heart.png.
+///
+/// Common prefixes will be removed, so the following would be a valid bundle
+/// layout for an asset group named "icons/heart":
+///
+/// icons/heart.meta
+/// icons/1.0x/heart.png
+/// icons/1.5x/heart.png
+/// icons/2.0x/heart.png
+class ResolutionAwareAssetResolver extends CandidateAssetResolver {
+  ResolutionAwareAssetResolver({ AssetBundle bundle, this.context })
+    : super(bundle: bundle);
+
+  BuildContext context;
```

Comment:
I'm confused by the lifetime/ownership model of this object. It's not private, so presumably it can be created at will and passed around. Its `context` isn't final, so presumably it's intended to be updated? By whom? But this isn't itself a Widget? How can you know when it should be disposed?

I think I'd take a pixel ratio rather than a build context and skip all the pain. :-) That would also mean that the Inherited update logic would work more reliably (since it would update the place that actually read the ratio, not the place that happened to have its context used when the ResolutionAwareAssetResolver was created).